src/python/algorithms/psolgent.py

```diff
@@ -218,31 +259,54 @@ def _get_vel(self):
                 (self.c2 * dot(u2, (self.global_best - self.pos))) +
                 (self.c3 * dot(u3, (self.local_best - self.pos))))
 
-    def _update_best(self, old, new):
+    # also implicitly updates self.best_path
+    def _update_best(self, old, new, old_rands, new_rands, old_paths, new_paths):
         # Updates the best objective function values for each member
-        old_fitness = array(self._get_fitness(old))
-        new_fitness = array(self._get_fitness(new))
-        best = zeros(shape=old.shape)
+        old_fitness = array(self._get_fitness(old_paths))
+        new_fitness = array(self._get_fitness(new_paths))
+        best = zeros(shape=old.shape)# * np.nan
+        best_rands = zeros(shape=old_rands.shape)# * np.nan
+        best_paths = zeros(shape=len(old_paths), dtype='object')
+        best_fitness = zeros(shape=len(old))
         if any(of < nf for of, nf in zip(old_fitness, new_fitness)):
             # replace indices in best with old members if lower fitness
             idx_old = [
                 idx for idx, val in enumerate(zip(old_fitness, new_fitness))
                 if val[0] < val[1]
             ]
             best[idx_old] = old[idx_old]
-            idx_new = where(best == 0)
+            best_rands[idx_old] = old_rands[idx_old]
+            best_paths[idx_old] = old_paths[idx_old]
+            best_fitness[idx_old] = old_fitness[idx_old]
+            # replace this with just find the rows where it is all zero
+            idx_new = np.all(best == 0, axis=1)
             # replace indices in best with new members if lower fitness
             best[idx_new] = new[idx_new]
+            best_rands[idx_new] = new_rands[idx_new]
+            best_paths[idx_new] = new_paths[idx_new]
+            best_fitness[idx_new] = new_fitness[idx_new]
         else:
             best = new
+            best_rands = new_rands
+            best_paths = new_paths
+            best_fitness = new_fitness
         self.best = array(best)
+        self.best_rands = array(best_rands)
+        self.best_paths = array(best_paths)
+        self.best_fitness = array(best_fitness)
+        # Update best known path if better
+        if not self.best_fit or self.best_fit > min(self.best_fitness):
+            self.best_path = list(self.best_paths[argmin(best_fitness)][:])
```

Comment:
Unfortunately this may lead to resource infeasible paths being saved here

Reply:
Ah, that's a good point.  There should be another check for if the path itself is feasible.  I think since I updated the `check_feasibility` method to optionally not save, we can do this without side effects here.